pieces/regex_json/rust/src/main.rs

```diff
@@ -20,13 +21,13 @@ struct CompanyFinal {
     annual_revenue_upper: f64,
 }
 
-fn get_data(path: &Path) -> Vec<Company> {
+fn get_data(path: &Path) -> Result<Vec<Company>> {
     if !path.exists() {
-        panic!("File {:?} not found", path);
+        bail!("File {:?} not found", path);
```

Comment:
Is there a benefit to using `bail` here over `panic`? It's not clear from the writeup if `bail` a like-for-like replacement for `panic` under all circumstances?

Reply:
There is an advantage. With `panic!` the program exits with no way to handle the error. `bail!` returns the result as an error with the message provided so it can be handled by the calling function.

Reply:
Sure, but in this case, isn't the right behaviour to exit without recovering if the file isn't found? Because that would mean that we don't have any data to proceed. There are definitely other situations in this file where bailing might be a better way, but for missing files there's an argument to be made that panicking is the "right" outcome.

Reply:
The problem is there is no option to do anything with it. Imagine it is a library you have installed that tries to read the file and panics. In this case the library creator has decided for you your program is going to crash, and they have decided what the message will be. 

Or is this was a web app a panic would crash your server. Instead you would probably want to send the error to a logger and return a message to the end user, but wouldn’t want the server to crash. 

Same with a desktop app/cli, you may want to exit the program, but you can do it in a way that shows the end user a nice message explaining what happened rather than crash output from rust.

Reply:
Yes, I can see this. I added more to the explanation.